corehq/apps/userreports/tasks.py

```diff
@@ -138,8 +146,70 @@ def rebuild_indicators_in_place(indicator_config_id, initiated_by=None, source=N
             config.meta.build.rebuilt_asynchronously = False
             config.save()
 
-        adapter.build_table(initiated_by=initiated_by, source=source)
-        _iteratively_build_table(config, in_place=True)
+        rows_count_before_rebuild = _get_rows_count_from_existing_table(adapter)
+        try:
+            adapter.build_table(initiated_by=initiated_by, source=source)
+            _iteratively_build_table(config, in_place=True)
+        except Exception:
+            _report_ucr_rebuild_metrics(config, source, 'rebuild_datasource_in_place', adapter,
+                                        rows_count_before_rebuild, error=True)
+            raise
+        _report_ucr_rebuild_metrics(config, source, 'rebuild_datasource_in_place', adapter,
+                                    rows_count_before_rebuild)
+
+
+def _get_rows_count_from_existing_table(adapter):
+    if not adapter.table_exists:
+        return None
+    table = adapter.get_existing_table_from_db()
+    return adapter.session_helper.Session.query(table).count()
+
+
+def _report_ucr_rebuild_metrics(config, source, action, adapter, rows_count_before_rebuild, error=False):
+    if source not in ('edit_data_source_rebuild', 'edit_data_source_build_in_place'):
+        return
+    try:
+        _report_metric_number_of_days_since_first_build(config, action)
+        if error:
+            _report_metric_rebuild_error(config, action)
+        else:
+            _report_metric_increase_in_rows_count(config, action, adapter, rows_count_before_rebuild)
+    except Exception:
+        pass
+
+
+def _report_metric_number_of_days_since_first_build(config, action):
+    try:
+        earliest_entry = DataSourceActionLog.objects.filter(
+            domain=config.domain,
+            indicator_config_id=config.get_id,
+            action__in=[DataSourceActionLog.BUILD, DataSourceActionLog.REBUILD]
+        ).earliest('date_created')
+    except DataSourceActionLog.DoesNotExist:
+        pass
+    else:
+        no_of_days = (datetime.utcnow() - earliest_entry.date_created).days
+        metrics_gauge(f'commcare.ucr.{action}.days_since_first_build', no_of_days, tags={'domain': config.domain})
+
+
+def _report_metric_rebuild_error(config, action):
+    from corehq.apps.userreports.views import _number_of_records_to_be_iterated_for_rebuild
+    expected_rows_to_process = _number_of_records_to_be_iterated_for_rebuild(config)
+    metrics_gauge(
+        f'commcare.ucr.{action}.failed.expected_rows_to_process',
+        expected_rows_to_process,
+        tags={'domain': config.domain}
+    )
+
+
+def _report_metric_increase_in_rows_count(config, action, adapter, rows_count_before_rebuild):
+    if rows_count_before_rebuild is None:
+        return
+    # Row count can only be obtained for synchronous rebuilds.
+    if not config.asynchronous:
+        rows_count_after_rebuild = adapter.get_query_object().count()
```

Comment:
nit: minor but this would "expected" count after rebuild, though that count eventually could be higher or lower than this number here.

Reply:
My understanding is that it would be the actual count as we are fetching the rows count from the database.  Also testing confirms this.

Reply:
Are we calling this after the rebuild has finished?
It does seem that this is being called at times directly after queuing the rebuild task.

Is this the count from the table or the count from the query the table will perform?

Reply:
> Are we calling this after the rebuild has finished?
> It does seem that this is being called at times directly after queuing the rebuild task.

Do you mean the `get_query_object().count()` method ?  If yes, it is used at multiple places and may be called before the rebuild.

For this code, this particular method is called by `_report_ucr_rebuild_metrics` which is used [here](https://github.com/dimagi/commcare-hq/blob/486494417e94dd16a38ed9905d0957e5cf50e227/corehq/apps/userreports/tasks.py#L157) at the end after rebuild code is finished. 

> Is this the count from the table or the count from the query the table will perform?

It is the count from the table.

Reply:
> Do you mean the get_query_object().count() method ? If yes, it is used at multiple places and may be called before the rebuild.

So if its run before the rebuild, the table would have 0 rows or old rows?

Reply:
> So if its run before the rebuild, the table would have 0 rows or old rows?

Good question. You are right.

So there are two scenarios here:

One good example is to to Preview Data where this is called.
1. Table has never been rebuilt :  In that case, it should an appropriate message however today it throws programming error (500 error) because no table exists. . I believe we have a ticket planned to address this 500 error. 

2. Table has been rebuilt before : Here the old table rows count.

Reply:
Right, so won't the metric be reported wrong in that case?

Reply:
> Right, so won't the metric be reported wrong in that case?

Apologies. I did not state clearly above. 

'For this code, this particular method is called by _report_ucr_rebuild_metrics which is used [here](https://github.com/dimagi/commcare-hq/blob/486494417e94dd16a38ed9905d0957e5cf50e227/corehq/apps/userreports/tasks.py#L157) at the end after rebuild code is finished.'

The metric code (as started above in the [earlier comment](https://github.com/dimagi/commcare-hq/pull/35496#discussion_r1883787624)) is only called inside the rebuild task at the end when the rebuild operation finishes.

Reply:
okay